onnxruntime/core/session/inference_session.cc

```diff
@@ -113,16 +113,68 @@ inline std::basic_string<T> GetCurrentTimeString() {
 
 #if !defined(ORT_MINIMAL_BUILD)
 
-bool AreAllNodesInMainGraphAssignedToOneEp(const Graph& graph, ProviderType provider) {
+/* This method returns ture if all *compute* nodes are placed on the CUDA EP
+   and all shape nodes are placed on the CPU EP
+ */
+std::pair<bool, int> AreAllComputeNodesAssignedToCudaEp(const Graph& graph) {
+  InlinedHashSet<NodeIndex> shape_nodes;
+  InlinedHashSet<NodeIndex> bfs_visited;
+  std::queue<NodeIndex> bfs_queue;
+
+  // Perform BFS to collect all nodes between all Shape -> Reshape node pairs
+  for (const auto& node : graph.Nodes()) {
+    if (node.OpType() == "Shape") {
+      for (auto iter = node.OutputNodesBegin(), end = node.OutputNodesEnd(); iter != end; ++iter) {
+        // If we haven't picked this node for BFS processing yet, then pick it up
+        if ((bfs_visited.find(iter->Index()) == bfs_visited.end())) {
+          bfs_visited.insert(iter->Index());
+          bfs_queue.push(iter->Index());
+        }
+      }
+    }
+  }
+
+  while (!bfs_queue.empty()) {
+    auto node_index = bfs_queue.front();
+    bfs_queue.pop();
+
+    const auto* node = graph.GetNode(node_index);
+
+    shape_nodes.insert(node_index);
+
+    for (auto iter = node->OutputNodesBegin(), end = node->OutputNodesEnd(); iter != end; ++iter) {
+      // If the child is not a Reshape node and we haven't processed/visited the node already,
+      // add the node for further processing
+      if (iter->OpType() != "Reshape" && (bfs_visited.find(iter->Index()) == bfs_visited.end())) {
+        bfs_visited.insert(iter->Index());
+        bfs_queue.push(iter->Index());
+      }
+    }
+  }
+
   for (const auto& node : graph.Nodes()) {
     const auto& node_provider = node.GetExecutionProviderType();
 
-    if (node_provider.empty() || node_provider != provider) {
-      return false;
+    // This node is not partitioned to the CUDA EP
+    // Ensure that this is a shape node assigned to the CPU EP
+    if (node_provider != onnxruntime::kCudaExecutionProvider) {
+      // If this node is partitioned to any other EP other
+      // than the CPU EP, then this isn't a shape node.
+      // Empty provider string means CPU EP.
+      if (!node_provider.empty() &&
+          node_provider != onnxruntime::kCpuExecutionProvider) {
+        return std::make_pair(false, -1);
+      }
+
+      // Check if this node is a shape node - If it isn't then we
+      // have found a compute node assigned to the CPU EP
+      if (shape_nodes.find(node.Index()) == shape_nodes.end()) {
+        return std::make_pair(false, -1);
+      }
```

Comment:
It will be helpful if we could add a log here to show which computing nodes are not assigned to CUDA EP. It will provide more insights for users to understand why this onnx model could not work with CUDA Graph. Further the users could try some changes to avoid using these ops in the onnx model.

Reply:
That information is already available to users when they turn on verbose logging (node placements)

Reply:
Got it!